File: 12b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use constant DUPLICATE => 1;

use ARGV::OrDATA;

my %caves;
while (<>) {
    chomp;
    my ($from, $to) = split /-/;
    undef $caves{$from}{$to};
    undef $caves{$to}{$from};
}

my @paths = ([['start'], 0]);
while (my @extendable = grep $paths[$_][0]->[-1] ne 'end', 0 .. $#paths) {
    for my $ie (reverse @extendable) {
        my $e = splice @paths, $ie, 1;
        for my $next (keys %{ $caves{ $e->[0][-1] } }) {
            next if $next eq 'start';

            my @new = (@{ $e->[0] }, $next);
            if ($next =~ /^[[:upper:]]/o) {
                push @paths, [\@new, $e->[DUPLICATE]];

            } elsif ($e->[DUPLICATE]) {
                push @paths, [\@new, 1]
                    unless grep $next eq $_, @{ $e->[0] };

            } else {
                push @paths, [\@new, 0+!! grep $next eq $_, @{ $e->[0] }];
            }
        }
    }
}
say scalar @paths;

__END__
start-A
start-b
A-c
A-b
b-d
A-end
b-end

File: 12b.py
#! /usr/bin/python3
import sys
import time

DUPLICATE = 1

caves = {}
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        (From, To) = line.rstrip("\n").split('-')
        if From in caves:
            caves[From].add(To)
        else:
            caves[From] = {To}
        if To in caves:
            caves[To].add(From)
        else:
            caves[To] = {From}

paths = [[['start'], False]]
while True:
    extendable = list(reversed(list(filter(lambda i:paths[i][0][-1] != 'end',
                                           range(len(paths))))))
    if not extendable:
        break

    for ie in extendable:
        e = paths.pop(ie)

        for nxt in caves[ e[0][-1] ]:
            if 'start' == nxt:
                continue

            new = e[0].copy()
            new.append(nxt)
            if nxt.upper() == nxt:
                paths.append([new, e[DUPLICATE]])
            elif e[DUPLICATE]:
                if not nxt in e[0]:
                    paths.append([new, True])
            else:
                paths.append([new, nxt in e[0]])

print(len(paths))
